File: LeetCode/remote/contains_duplicates_III.rb
# Given an array of integers, find out whether there are two distinct indices i and j in the array such that the difference between nums[i] and nums[j] is at most t and the difference between i and j is at most k. 
#

require 'test/unit'
extend Test::Unit::Assertions

class Deque
    def initialize(flag)
        @data = []
        @flag = flag
    end

    def compare(x, y)
        @flag ? [x, y].max : [x, y].min
    end

    def enqueue(value)
        count = 0
        while @data.any? && compare(@data.last[0], value) == value
            count += @data.last[1] + 1
            @data.pop
        end

        @data.push([value, count])
    end

    def dequeue
        if @data.first[1] > 0
            @data.first[1] -= 1
        end

        @data.shift 
    end

    def head
        return @data.first[0] unless @data.first.nil?
    end
end

def contains_nearby_almost_duplicate(nums, k, t)
    dequeue_max = Deque.new(true)
    dequeue_min = Deque.new(false)
    return false if k == 0
    nums.each_with_index { |x, i|
        dequeue_max.dequeue if i > k
        dequeue_min.dequeue if i > k 
        puts dequeue_min.inspect
        puts dequeue_max.inspect
        return true if !dequeue_max.head.nil? && dequeue_max.head == x ||
                       !dequeue_min.head.nil? && dequeue_min.head == x
        return true if !dequeue_max.head.nil? && x < dequeue_max.head && dequeue_max.head - x <= t || 
                       !dequeue_min.head.nil? && x > dequeue_min.head && x - dequeue_min.head <= t
        dequeue_max.enqueue(x)
        dequeue_min.enqueue(x)
    }

    return false
end

assert_equal(true, contains_nearby_almost_duplicate([1, 2, 3, 4, 5], 2, 2))
assert_equal(true, contains_nearby_almost_duplicate([-1, -1], 1, 0))
assert_equal(false, contains_nearby_almost_duplicate([1, 2], 0, 1))
assert_equal(true, contains_nearby_almost_duplicate([2, 1], 1, 1))
assert_equal(false, contains_nearby_almost_duplicate([1, 3, 1], 1, 1))
assert_equal(true, contains_nearby_almost_duplicate([7, 1, 3], 2, 3))
assert_equal(true, contains_nearby_almost_duplicate([3, 6, 0, 2], 2, 2))

File: LeetCode/remote/contains_duplicates_III.cs
//  https://leetcode.com/problems/contains-duplicate-iii/
//
//  Given an array of integers,
//  find out whether there are two distinct indices i and j in the array such that the difference
//  between nums[i] and nums[j] is at most t and the difference between i and j is at most k. 

using System;
using System.Collections.Generic;
using System.Linq;

public class Solution {
    class Dequeue
    {
        private LinkedList<int> storage = new LinkedList<int>();
        
        public Dequeue()
        {
            Empty = true;
        }
        
        public void Add(int x, Func<int, int, bool> comparer)
        {
            while (storage.Count > 0 && comparer(storage.Last(), x))
            {
                storage.RemoveLast();
            }
            
            storage.AddLast(x);
            Empty = false;
        }
        
        public void Remove(int x)
        {
            if (storage.Count > 0 && storage.First() == x)
            {
                storage.RemoveFirst();
            }
            
            Empty = storage.Count == 0;
        }
        
        public int Head {
            get {
                return storage.First();
            }
        }
        
        public bool Empty { 
            get; 
            private set; 
        }
    }
    
    public bool ContainsNearbyAlmostDuplicate(int[] nums, int k, int t) {
        var max = new Dequeue();
        var min = new Dequeue();
        for (var i = 0; i < Math.Min(k, nums.Length); i++)
        {
            max.Add(nums[i], (a, b) => a >= b);
            min.Add(nums[i], (a, b) => a < b);
        }

        if (!max.Empty && !min.Empty && max.Head - min.Head <= t)
        {
            return true;
        }
        
        for (var i = k; i < nums.Length; i++)
        {
            max.Add(nums[i], (a, b) => a >= b);
            min.Add(nums[i], (a, b) => a < b);
            max.Remove(nums[i - k]);
            min.Remove(nums[i - k]);
            
            if (!max.Empty && !min.Empty && max.Head - min.Head <= t)
            {
                return true;
            }
        }
        
        return false;
    }

    public bool ContainsNearbyAlmostDuplicate(int[] nums, int k, int t) {
        var set = new SortedDictionary<int, int>();

        for (var i = 0; i < nums.Length; i++)
        {
            if (set.Any() && set.Keys.Last() - set.Keys.First() <= t)
            {
                return true;
            }

            if (!set.ContainsKey(nums[i]))
            {
                set[nums[i]] = 0;
            }

            set[nums[i]]++;
            if (i >= k) 
            {
                set[nums[i - k]]--;

                if (set[nums[i - k]] == 0)
                {
                    set.Remove(nums[i - k]);
                }
            }
        }

        return false;
    }

    static void Main()
    {
        Console.WriteLine(new Solution().ContainsNearbyAlmostDuplicate(new [] { 1, 2, 3, 4, 5, 6 }, 4, 3 ));
        Console.WriteLine(new Solution().ContainsNearbyAlmostDuplicate(new [] { 1, 2, 3, 4, 5, 6 }, 4, -1 ));
        Console.WriteLine(new Solution().ContainsNearbyAlmostDuplicate(new [] { 0 }, 0, 0 ));
        Console.WriteLine(new Solution().ContainsNearbyAlmostDuplicate(new [] { -3, 3 }, 2, 4 ));
    }
}
